## Basic input and output

This text discusses basic input and output operations in Python. It begins by introducing a simple "Hello, world" program that can be executed in a Python environment. Users are encouraged to edit the text within quotes and rerun the program.

The text also demonstrates how to print multiple strings, which are concatenated by default with spaces. It explains that numerical expressions within the print function are evaluated and converted to strings, and then concatenated with spaces.

Additionally, the text introduces the input function, which allows users to read textual input from the user. This function is given a string parameter, which is printed as a prompt to the user. The example provided shows how to use the input function to store a user's name in a variable and then print a greeting with that name.

In summary, the text covers basic Python concepts related to printing, user input, and program execution.

In [1]:
print("Hello, world!")

Hello, world!


In [2]:
print("Hello", "John!", "How are you?")

Hello John! How are you?


In [3]:
print(1, "plus", 2, "equals", 1+2)

1 plus 2 equals 3


In [4]:
name=input("Give me your name: ")
print("Hello,", name)

Hello, John


## Indentation
- For Loops in Python

Repetition is possible with the for loop in Python.
The body of the for loop is indented with a tabulator or four spaces.

Unlike some other languages, Python doesn't use braces to denote the body of the loop.
When the indentation stops, the body of the loop ends.

Example:

```python

for i in range(3):
    print("Hello")
print("Bye!")
```
Output:
Hello
Hello
Hello
Bye!

Indentation in Python:

Indentation is a crucial part of Python syntax.
It is used not only in for loops but also in other compound statements, such as function bodies, different branches of if statements, and while loops.

- The range() Function:
The range(3) expression results in a sequence of integers: 0, 1, and 2.
In Python, range represents a half-open interval, with the end point excluded from the range.
In general, range(n) gives integers from 0 up to (n-1).

- Printing Variable Values in a For Loop

You can modify the code to print the value of the variable i at each iteration of the loop.
This is done by adding print(i) inside the for loop.

- Modified Example:

```python
for i in range(3):
    print("Hello")
    print(i)  # Added to print the value of i
print("Bye!")
```
When you run this modified code, it will print the value of i along with "Hello" and "Bye!" at the end.

## Exercise 1.1 (Hello world)

Fill in the missing piece in the solution stub file hello_world.py in folder src to make it print the following:

Hello, world!
Make sure you use correct indenting. You can run it with command python3 src/hello_world.py. If the output looks good, then you can test it with command tmc test. If the tests pass, submit your solution to the server with command tmc submit.

In [6]:
print("Hello, world!")

Hello, world!


## Exercise 1.2 (compliment)
Fill in the stub solution to make the program work as follows. The program should ask the user for an input, and then print an answer as the examples below show.

What country are you from? Sweden
I have heard that Sweden is a beautiful country.

What country are you from? Chile  
I have heard that Chile is a beautiful country.

In [7]:
country = input("What country are you from? ")
print(f"I have heard that {country} is a beautiful country.")

I have heard that China is a beautiful country.


## Exercise 1.3 (multiplication)
Make a program that gives the following output. You should use a for loop in your solution.

4 multiplied by 0 is 0

In [9]:
number = 4
for first in range(11):
    product = first * number
    print(f"{number} multiplied by {first} is {product}")

4 multiplied by 0 is 0
4 multiplied by 1 is 4
4 multiplied by 2 is 8
4 multiplied by 3 is 12
4 multiplied by 4 is 16
4 multiplied by 5 is 20
4 multiplied by 6 is 24
4 multiplied by 7 is 28
4 multiplied by 8 is 32
4 multiplied by 9 is 36
4 multiplied by 10 is 40


## Variables and data types

Python allows simple variable assignment, e.g., a = 1, without explicitly specifying the variable type.

You can determine a variable's type using the type function.

Python uses dynamic typing, meaning the type of a variable is based on its value, not its name.

Variables are names that refer to values, and the assignment operator binds the name to the value.

Basic data types in Python include int, float, complex, str, bool, and bytes.

Examples of data type usage: i = 5, f = 1.5, b = i == 4, c = 0 + 2j, s = "conca" + "tenation".

Type names can be used for type conversion, e.g., int(-2.8), float(2), int("123"), bool(-2), bool(0), str(234).

Bytes are used to represent information with values between 0 and 255, consisting of 8 bits, commonly used in storage and data transmission.

Characters are encoded as sequences of bytes, and you can encode and decode them using methods like .encode() and .decode().

Understanding bytes may be necessary when dealing with specific data sets and specifying character encoding

In [10]:
a = 1
print(a)

1


In [11]:
type(a)

int

In [12]:
a = "some text"
type(a)

str

In [13]:
i = 5
f = 1.5
b = i == 4
print("Result of the comparison:", b)
c=0+2j # Note that j denotes the imaginary unit of complex numbers.
print("Complex multiplication:", c * c)
s="conca" + "tenation"
print(s)

Result of the comparison: False
Complex multiplication: (-4+0j)
concatenation


In [14]:
print(int(-2.8))
print(float(2))
print(int("123"))
print(bool(-2), bool(0)) # Zero is interpreted as False
print(str(234))


-2
2.0
123
True False
234


In [16]:
b = "ä".encode("utf-8") # Convert characters(s) to a sequence of bytes

print(b) # Prints bytes in hexadecimal notation

print(list(b))  # Prints bytes in decimal notation


b'\xc3\xa4'
[195, 164]


In [18]:
bytes.decode(b, "utf-8") # convert sequence of bytes to character(s)

'ä'

## Creating strings
A string is a sequence of characters enclosed in single ('') or double ("") quotes, allowing for flexibility when dealing with quotation marks inside a string.

Strings can contain escape sequences like '\n' for a newline and '\t' for a tab.

Triple quotes (''' or """) are used to create multiline strings.

Concatenating strings with the '+' operator is less efficient for many strings. The 'join' method is recommended for efficiency.

String interpolation is a way to insert values into strings. It offers more readability and is available through various methods, including Python format strings, the format method, and f-strings.

Examples of string interpolation methods: % for Python format strings, format() method, and f-strings (f"...").

Format specifiers (e.g., %i, %f, :d, .1f) allow you to control the formatting of values within the string.

String interpolation is a matter of personal preference; commonly used methods are f-strings and the format method.

Overall, string interpolation provides a cleaner and more efficient way to embed values within strings.

In [19]:
print("One\tTwo\nThree\tFour")

One	Two
Three	Four


In [20]:
s="""A string
spanning over
several lines"""

In [21]:
a="first"
b="second"
print(a+b)
print(" ".join([a, b, b, a])) # More about the join method later

firstsecond
first second second first


In [22]:
print(str(1) + " plus " + str(3) + " is equal to " + str(4))
# slightly better
print(1, "plus", 3, "is equal to", 4)

1 plus 3 is equal to 4
1 plus 3 is equal to 4


In [23]:
print("%i plus %i is equal to %i" % (1, 3, 4)) # Format syntax

print("{} plus {} is equal to {}".format(1, 3, 4)) # Format method

print(f"{1} plus {3} is equal to {4}") # f-string

1 plus 3 is equal to 4
1 plus 3 is equal to 4
1 plus 3 is equal to 4


In [24]:
print("%.1f %.2f %.3f" % (1.6, 1.7, 1.8)) # Old style 

print("{:.1f} {:.2f} {:.3f}".format(1.6, 1.7, 1.8)) # newer style

print(f"{1.6:.1f} {1.7:.2f} {1.8:.3f}") # f-string

1.6 1.70 1.800
1.6 1.70 1.800
1.6 1.70 1.800


In [26]:
print("%s concatenated with %s produces %s" % ("water", "melon", "water"+"melon"))
print("{0} concatenated with {1} produces {0}{1}".format("water", "melon"))
print(f"{'water'} concatenated with {'melon'} produces {'water' + 'melon'}")

water concatenated with melon produces watermelon
water concatenated with melon produces watermelon
water concatenated with melon produces watermelon


## Expressions

Expressions:

An expression in Python is a piece of code that evaluates to a value. It combines values (literals or variables) with operators.

Values in expressions can be literals (e.g., numbers, strings) or variables.

Operators used in expressions include arithmetic operators (+, -, *, /), comparison operators (<, >, ==), function calls (e.g., cos(0)), indexing (e.g., mylist[1]), and attribute references (e.g., obj.attr).

Examples of expressions include mathematical calculations (e.g., 1+2, 7/(2+0.1), variable references (e.g., a), and logical expressions (e.g., c > 0 and c != 1).

Expressions are used to perform calculations, generate results, and can be part of statements or other expressions.

Expressions are fundamental building blocks in Python programming.

Statements:

Statements are commands in Python that perform some action or have an effect. They include variable assignments, function calls (not part of an expression), and control flow statements.

Variable assignments, like i = 5, assign a value to a variable. You can increment variables using i = i + 1 or use the shorthand i += 1.

Python lacks ++ and -- operators seen in some other languages.

Augmented assignment operators, such as +=, -= are available for the arithmetic operators, making operations like incrementing simpler.

Flow control statements like if-else, for loops, and while loops are used to control the flow of a program and execute code conditionally or iteratively.

Statements are essential for creating structured and meaningful programs.

Understanding the distinction between expressions and statements is important for writing Python code and controlling program execution. Expressions evaluate to values, while statements are commands that perform actions.


## For loop and While loop

A for loop is used to execute a block of statements a specific number of times or for each item in an iterable.

In a for loop, the variable (commonly named i or any name you choose) represents the current item from the iterable on each iteration.

You can use a for loop with a list, generator, or any iterable object to iterate through its elements.

The range() function generates a sequence of values from 0 to n-1, making it a common choice for simple iteration.

In Python, for loops are suitable for iterating through elements in an iterable, such as lists, dictionaries, strings, and more.

While loops are an alternative when you need to iterate based on a condition or until a specific condition is met, making them useful for cases like generating Fibonacci numbers within a limit.

Python provides a clean and concise way to work with for loops, making them a natural choice for many iteration tasks.

Iterables and generators are concepts that are closely related to the use of for loops, and they provide a way to work with sequences of data. These will be discussed in more detail in later lessons.

For loops are a powerful tool for iterating through sequences, making them a fundamental part of Python programming. Depending on the specific use case, you can choose between for loops and while loops to achieve your desired iteration behavior.

In [27]:
i = 1
while i * 1 < 1000:
    print("Square of", i, "is", i*i)
    i = i + 1
print("Finished printing all the squares below 1000.")

Square of 1 is 1
Square of 2 is 4
Square of 3 is 9
Square of 4 is 16
Square of 5 is 25
Square of 6 is 36
Square of 7 is 49
Square of 8 is 64
Square of 9 is 81
Square of 10 is 100
Square of 11 is 121
Square of 12 is 144
Square of 13 is 169
Square of 14 is 196
Square of 15 is 225
Square of 16 is 256
Square of 17 is 289
Square of 18 is 324
Square of 19 is 361
Square of 20 is 400
Square of 21 is 441
Square of 22 is 484
Square of 23 is 529
Square of 24 is 576
Square of 25 is 625
Square of 26 is 676
Square of 27 is 729
Square of 28 is 784
Square of 29 is 841
Square of 30 is 900
Square of 31 is 961
Square of 32 is 1024
Square of 33 is 1089
Square of 34 is 1156
Square of 35 is 1225
Square of 36 is 1296
Square of 37 is 1369
Square of 38 is 1444
Square of 39 is 1521
Square of 40 is 1600
Square of 41 is 1681
Square of 42 is 1764
Square of 43 is 1849
Square of 44 is 1936
Square of 45 is 2025
Square of 46 is 2116
Square of 47 is 2209
Square of 48 is 2304
Square of 49 is 2401
Square of 50 is 2500
Sq

In [29]:
s = 0
for i in [0,1,2,3,4,5,6,7,8,9]:
    s = s + i
print("The sum is", s)

The sum is 45


## Exercise 1.4 (multiplication table)

- Create a multiplication table using two nested for loops, where the numbers are aligned with a field width of four.
- Each entry in the table represents the product of the row and column number.

- Note:
- print("text", end="")
- print("more text")

In [34]:
for rowNumber in range(1, 11):
    for columnNumber in range(1, 11):
        product = rowNumber * columnNumber
        # formatting the product variable with a width of four characters
        print(f"{product:4}", end="")

    print()

   1   2   3   4   5   6   7   8   9  10
   2   4   6   8  10  12  14  16  18  20
   3   6   9  12  15  18  21  24  27  30
   4   8  12  16  20  24  28  32  36  40
   5  10  15  20  25  30  35  40  45  50
   6  12  18  24  30  36  42  48  54  60
   7  14  21  28  35  42  49  56  63  70
   8  16  24  32  40  48  56  64  72  80
   9  18  27  36  45  54  63  72  81  90
  10  20  30  40  50  60  70  80  90 100


## Descision making with the if statment and loop
1. Using the if statement:

In the first example, the user is asked to input an integer.
The code checks whether the input integer x is greater than or equal to 0.
If x is non-negative, it assigns x to the variable a. If x is negative, it assigns the absolute value of x to a.
Finally, it prints the absolute value of x.
2. The general form of an if-else statement:

This section provides a general structure for if-else statements, which allows you to handle multiple conditions and execute different code blocks based on those conditions.
3. Another example of if-else statement:

In this example, the user is asked to input a floating-point number.
The code checks whether the input number c is positive, negative, or zero, and prints a corresponding message.
4. Breaking the loop with break:

This code demonstrates how to use the break statement to exit a loop when a specific condition is met.
It iterates through a list l and breaks out of the loop as soon as it encounters the first negative element.
The code then prints the first negative element found.
5. Continuing the loop with continue:

This example shows how to use the continue statement to skip the current iteration of a loop and move to the next one.
It iterates through the list l and, when it encounters a negative number, it continues to the next iteration.
For non-negative numbers, it calculates and prints the square root and natural logarithm.
These code snippets provide a practical understanding of how to use if-else statements and control flow in Python, including the use of break and continue in loops.

In [35]:
x = input("Give an integer: ")
x = int(x)
if x >= 0:
    z=x
else:
    a=-x
print("The absolute value of %i is %i" % (x, a))

The absolute value of -1 is 1


In [36]:
# second if statment example
c=float(input("Give a number: "))
if c > 0:
    print("c is positive")
elif c < 0:
    print("c is negative")
else:
    print("c is zero")

c is positive


In [37]:
# for loop in array
l=[1, 3, 65, 3, -1, 56, -10]
for x in l:
    if x < 0:
        break

print("The first negative list element was", x)

The first negative list element was -1


In [39]:
from math import sqrt, log
l = [1, 3, 65, 3, -1, 56, -10]
for x in l:
    if x < 0:
        continue
    print(f"Square root of {x} is {sqrt(x):.3f}")
    print(f"Natural logarithm of {x} is {log(x):.4f}")

Square root of 1 is 1.000
Natural logarithm of 1 is 0.0000
Square root of 3 is 1.732
Natural logarithm of 3 is 1.0986
Square root of 65 is 8.062
Natural logarithm of 65 is 4.1744
Square root of 3 is 1.732
Natural logarithm of 3 is 1.0986
Square root of 56 is 7.483
Natural logarithm of 56 is 4.0254


## Exercise 1.5(two dice)
- Let us consider throwing two dice. 
(A dice can give a value between 1 and 6.) 
- Use two nested for loops in the main function to iterate through all possible combinations the pair of dice can give. 
- There are 36 possible combinations. 
- Print all those combinations as (ordered) pairs that sum to 5. 
- For example, your printout should include the pair (2,3). Print one pair per line.


In [40]:
# print dice sum pair equal to 5
target = 5
for firstDie in range(1, 7):
    for secondDie in range(1, 7):
        if firstDie + secondDie == target:
            print(f"({firstDie}, {secondDie})")

(1, 4)
(2, 3)
(3, 2)
(4, 1)


## Functions
- Defining a Function:

Functions are defined using the def statement followed by the function name, parameter list, and a colon.

A docstring enclosed in triple quotes can be used to provide documentation for the function's purpose and usage.

- Function Parameters:

Function parameters are specified within the parentheses and may include zero or more parameters.

In the example, you can see a function named double(x) that takes one parameter, x.

Parameters can have default values. In the example, you see the degree=2 in the length function.

- Function Execution:

Functions are executed when they are called. In the examples, you can see function calls like double(4) and sum_of_squares([-2, 4, 5]).

Functions can return values using the return statement. In the double function, it returns x*2.

Argument Packing and Unpacking:

You can use * to pack arguments into a tuple, as seen in the sum_of_squares function. It allows you to pass an arbitrary number of positional arguments.

Argument unpacking is done using the * when calling a function, as shown with sum_of_squares(*lst).

- Named Arguments and Default Values:

Functions can have named arguments, which allow you to pass values in any order. Named arguments must come after positional arguments.

You can specify default parameter values in the function definition. Users can override these defaults when calling the function.

- Documentation:

Docstrings are used to document a function's purpose and usage, and you can access them using function.__doc__ or help(function).
Examples:

The examples provided illustrate different aspects of function definition, usage, and parameter handling.
Understanding how to define functions and work with their parameters is fundamental in Python programming. It allows you to create reusable and organized code for various tasks.


In [1]:
def double(x):
    "This function multiplies its argument by two."
    return x * 2
print(double(4), double(1.2), double("abc")) # It even happens to work for strings!

8 2.4 abcabc


In [3]:
print("The docstring is:", double.__doc__)
help(double) # Another way to access the docstring

The docstring is: This function multiplies its argument by two.
Help on function double in module __main__:

double(x)
    This function multiplies its argument by two.



In [4]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



In [6]:
def sum_of_squares(a, b):
    "Computes the sum of arguments squared"
    return a**2 + b**2
print(sum_of_squares(3, 4))

25


In [7]:
def sum_of_square_list(lst):
    "Computes the sum of squares of elements in the list given as parameter"
    s=0
    for x in lst:
        s += x**2
    return s
print(sum_of_square_list([-2]))
print(sum_of_square_list([-2,4,5]))

4
45


In [11]:
def sum_of_squares_pack(*t):
    "Computes the sum of squares of arbitrary number of arguments"
    s=0
    for x in t:
        s += x**2
    return s
print(sum_of_squares_pack(-2))
print(sum_of_squares_pack(-2,4,5))

4
45


In [12]:
lst = [1,5,8]
print("With list unpacked as arguments to the functions:", sum_of_squares_pack(*lst))

With list unpacked as arguments to the functions: 90


In [13]:
def named(a, b, c):
    print("First:", a, "Second:", b, "Third:", c)
named(5, c=7, b=8)

First: 5 Second: 8 Third: 7


In [15]:
print(1, 2, 3, end=' |', sep=' _*_ ')
print("first", "second", "third", end=' |', sep=' -*- ')

1 _*_ 2 _*_ 3 |first -*- second -*- third |

In [16]:
def length(*t, degree=2):
    """
    Computes the length of the vector given as parameter. 
    By default, it computes the Euclidean distance (degree==2)
    """
    s=0
    for x in t:
        s += abs(x)**degree
    return s**(1/degree)
print(length(-4, 3))
print(length(-4, 3, degree=3))


5.0
4.497941445275415


## Visibility
- Variable Scope:

Functions create local scopes where variables are local and not accessible outside.
Global variables are accessible from both local and global scopes.
- Variable Visibility:

Local variables can shadow global variables of the same name.
Global variables can be read in local scopes.
- Modifying Global Variables:

To modify a global variable from within a function, use the global statement.
- Nested Functions:

Python allows defining functions within other functions, creating nested scopes.
Inner functions can access variables from outer functions, but they can shadow them as well.
- Global and Nonlocal Statements:

The global statement rebinds a global variable.
The nonlocal statement refers to the nearest outer variable (not global) in nested functions.
Understanding variable scope and using global and nonlocal statements is important for proper variable management in Python

In [17]:
i=2 # global variable
def f():
    i=3 # this creates a new variable, it does not rebind the global i
    print(i)    # This will print 3 
f()
print(i)        # This will print 2

3
2


In [18]:
i = 2
def f():
    global i 
    i = 5 # rebhind the global i variable
    print(i) # This will print 5
f()
print(i)    # This will print 5  

5
5


In [19]:
def f(): # outer function
    b=2
    def g(): # inner function
        #nonlocal b # Without this nonlocal statement,
        b=3 # this will create a new local variable
        print(b)

    g()
    print(b)
f()

3
2


## Exercise 1.6(triple square)
Write two functions: 
triple and square. Function triple multiplies its parameter by three. 
Function square raises its parameter to the power of two. 
For example, we have equalities triple(5)==15 and square(5)==25.

- Part 1.
In the main function write a for loop that iterates through values 1 to 10, and for each value prints its triple and its square. The output should be as follows:

- triple(1)==3 square(1)==1
- triple(2)==6 square(2)==4
...
- Part 2.
Now modify this for loop so that it stops iteration when the square of a value is larger than the triple of the value, without printing anything in the last iteration.

Note that the test cases check that both functions triple and square are called exactly once per iteration.

In [28]:
def triple(number):
    return number * 3

def square(number):
    return number ** 2

def main():
    triple_product =0
    square_product = 0
    for number in range(1, 11):
        triple_product = triple(number)
        square_product = square(number)
        if square_product > triple_product:
            break
        print(f"triple({number}) == {triple_product} square({number}) == {square_product}")
main()

triple(1) == 3 square(1) == 1
triple(2) == 6 square(2) == 4
triple(3) == 9 square(3) == 9


## Exercise 1.7 (dareas of shapes)

Create a program that can compute the areas of three shapes, triangles, rectangles and circles, when their dimensions are given.

An endless loop should ask for which shape you want the area be calculated. An empty string as input will exit the loop. If the user gives a string that is none of the given shapes, the message “Unknown shape!” should be printed. Then it will ask for dimensions for that particular shape. When all the necessary dimensions are given, it prints the area, and starts the loop all over again. Use format specifier f for the area.

What happens if you give incorrect dimensions, like giving string "aa" as radius? You don't have to check for errors in the input.

In [8]:
def triangle_area(base, height):
    return (base * height) * (1/2)

def rectangle_area(length, height):
    return length * height;

def circle_area(radius):
    import math 
    return math.pi * (radius ** 2)
area = 0
while True :
    user_input = input("Choose a shape (triangle, rectangle, circle): ")
    area = 0
    if user_input == "":
        break
    elif user_input == "triangle":
        base = float(input("Give base of the triangle: "))
        height = float(input("Give height of the triangle: "))
        area = triangle_area(base, height)
    elif user_input == "rectangle":
        length = float(input("Give width of the rectangle: "))
        width = float(input("Give height of the rectangle: "))
        area = rectangle_area(length, width)
    elif user_input == "circle":
        radius = float(input("Give radius of the circle: "))
        area = circle_area(radius)
    else:    
        print("Unknown shape!")
    if area > 0:
        print(f"The area is {area:.6f}")   
   


The area is 80.000000


## Main Data structures:

- Python's main data structures include strings, lists, tuples, dictionaries, and sets.
### Lists:

- Lists are flexible containers for storing an arbitrary number of elements, even zero.
- Elements are stored in sequential order, separated by commas, and enclosed in square brackets: [2, 100, "hello", 1.0].
- Lists can contain elements of different types.
### Tuples:

- Tuples are fixed-length, immutable, and ordered containers.
- Elements in tuples are separated by commas and enclosed in parentheses: (3,), (1,3), (1, "hello", 1.0).
- Tuples can also contain elements of different types.
### Sequences:

- Lists, tuples, and strings are referred to as sequences.
- Common characteristics among sequences include:
    - Length can be queried using len(sequence).
    - min() and max() functions find the minimum and maximum elements.
    - sum() adds all numeric elements together.
    - Sequences can be concatenated using the + operator.
    - Sequences can be repeated using the * operator (e.g., "hi" * 3 == "hihihi").
### Indexing and Slicing:

- Sequences are ordered and allow access to elements through indexing.
- Indexing starts from 0, and negative integers index from the end.
    - E.g., "abcd"[2] == "c", -1 refers to the last element.
- Slicing allows you to extract a subsequence using the range [start:stop].
    - The stop index is exclusive.
    - Generic slicing: sequence[start:stop:step].
    - Default values: start=0, stop=len(sequence), step=1.
    - For example, "abcde"[1:] == "bcde" and [0,1,2,3,4,5,6,7,8,9][::3] returns [0, 3, 6, 9].
    Understanding these data structures and their common operations is essential for effective Python programming.

## Exercise 1.8 (Solve quadratic)
Formula:
ax^2 + bx + c = 0

Write a function solve_quadratic, that returns both solutions of a generic quadratic as a pair (2-tuple) when the coefficients are given as parameters. It should work like this:

print(solve_quadratic(1,-3,2))
(2.0,1.0)
print(solve_quadratic(1,2,1))
(-1.0,-1.0)

You may want to use the math.sqrt function from the math module in your solution. Test that your function works in the main function!

In [9]:
def solve_quadratic(a, b, c):
    import math
    plus_value = 0
    subtract_value = 0
    denominator = (2 * a)
    result = (plus_value, subtract_value)
    sqrt_value = math.sqrt(b ** 2 - (4*a*c))
    b *= -1
    plus_value = (b + sqrt_value) / denominator
    subtract_value = (b - sqrt_value) / denominator

    result = (plus_value, subtract_value)
    return result
print(solve_quadratic(1, -3, 2))

(2.0, 1.0)


## Modifying Lists:

- Lists can be modified by assigning values to elements using indexing or slicing. For example, L[2] = 10 changes the third element of the list.

- You can assign a list to a slice to modify a portion of the list. For instance, L[1:3] = [4] replaces elements at positions 1 and 2 with the value 4.

- Lists can also be modified using list methods like append, extend, insert, remove, pop, reverse, and sort.

### Generating Numerical Sequences:

- The range function is used to generate numerical sequences automatically. It creates a sequence of numbers from 0 up to, but not including, the specified end value. For example, range(7) generates the sequence [0, 1, 2, 3, 4, 5, 6].

- The result of the range function is not a list but a sequence, similar to slices. You can access its elements or convert it to a list using the list constructor.

- The range function can take optional parameters to specify the start, end, and step of the sequence. For instance, range(0, 7, 2) generates the sequence [0, 2, 4, 6].

### Sorting Sequences:

- Python provides two ways to sort sequences. The sort method modifies the original list in-place, while the sorted function returns a new sorted list and leaves the original unchanged.

- The sorted function can take the reverse=True parameter to sort the sequence in descending order.

Understanding these concepts is essential for manipulating lists and sequences efficiently in Python.


## Exercise 1.9 (merge)

- Define a function named "merge" that takes two sorted lists, L1 and L2, as parameters.

- Initialize an empty list, "result," to store the merged list.

- Create two variables, "i" and "j," and set them to 0. These variables will be used to keep track of the current positions in L1 and L2, respectively.

- Create a while loop that continues as long as "i" is less than the length of L1 and "j" is less than the length of L2. This loop will help merge the elements from L1 and L2.

- Inside the loop, compare the elements at L1[i] and L2[j].

- Append the smaller of the two elements to the "result" list and increment the corresponding index (i or j).

- After exiting the while loop, there may be remaining elements in either L1 or L2. Check if "i" is less than the length of L1 or "j" is less than the length of L2. If either condition is true, it means there are remaining elements in one of the lists.

- If there are remaining elements in L1, extend the "result" list with the remaining elements in L1 starting from index "i."

- If there are remaining elements in L2, extend the "result" list with the remaining elements in L2 starting from index "j."

- Return the "result" list, which is now a sorted list containing all the elements from L1 and L2.

- In the main function, create two sorted lists, L1 and L2, using the "sorted" function or by manually sorting them.

- Call the "merge" function with L1 and L2 as arguments, and store the returned merged list in a variable.

- Verify that the merged list is sorted and has a length equal to the sum of the lengths of L1 and L2.

- Test the "merge" function with a couple of example inputs to ensure it works correctly.

In [13]:
def merge(L1, L2):
    result = []
    i = 0
    j = 0
    while i < len(L1) and j < len(L2):
        # merge data from L1 to L2
        if L1[i] < L2[j]:
            result.append(L1[i])
            i += 1
        else:
            result.append(L2[j])
            j += 1
    if i < len(L1):
        result.extend(L1[i:])
    elif j < len(L2):
        result.extend(L2[j:])
         
    return result
def main():
    L1 = [2,3,7,8]
    L2 = [9,18,4,6]
    L1.sort()
    L2.sort()
    L = merge(L1, L2)
    print(L)
    if len(L) == len(L1) + len(L2):
        print("sorted")
    else:
        print("not sorted")
main()

[2, 3, 4, 6, 7, 8, 9, 18]
sorted


## Exercise 1.10 (detect ranges)

Create a function named detect_ranges that gets a list of integers as a parameter. The function should then sort this list, and transform the list into another list where pairs are used for all the detected intervals. So 3,4,5,6 is replaced by the pair (3,7). Numbers that are not part of any interval result just single numbers. The resulting list consists of these numbers and pairs, separated by commas. An example of how this function works:

print(detect_ranges([2,5,4,8,12,6,7,10,13]))
[2,(4,9),10,(12,14)]
Note that the second element of the pair does not belong to the range. This is consistent with the way Python's range function works. You may assume that no element in the input list appears multiple times.

In [5]:
def detect_ranges(L):
    # sort the list
    # 2, 4, 5, 6, 7, 8, 10, 12, 13
    nums = sorted(L)
    start = end = nums[0]
    result = []
    # check the range of sorted array
    for num in nums[1:]:
        if num == end + 1:
            end = num
        else:
            # detect if we are getting a single digit
            if start == end:
                result.append(start)
            else: # detect if we are in a range
                result.append((start, end + 1)) 
            start = end = num
    
    # if we have a single digit without range then append that
    if start == end:
        result.append(start)
    else: # if we have a final range append that in a tuple form
        result.append((start, end + 1))
    return result 

print(detect_ranges([2,5,4,8,12,6,7,10,13]))

[2, (4, 9), 10, (12, 14)]


## Ziping sequences
The zip function combines two (or more) sequences into one sequence. If, for example, two sequences are zipped together, the resulting sequence contains pairs. In general, if n sequences are zipped together, the elements of the resulting sequence contains n-tuples.


In [6]:
L1 = [1, 2, 3]
L2 = ["first", "second", "third"]
print(zip(L1, L2)) # Note that zip does not return a list, like range
print(list(zip(L1, L2))) # Convert to a list

[(1, 'first'), (2, 'second'), (3, 'third')]


In [9]:
days = "Monday Tuesday Wednesday Thursday Friday Saturday Sunday".split()
weathers = "rainy rainy sunny cloudy rainy sunny sunny".split()
temperatures = [10, 12, 12, 9, 9, 11, 11]
for day, weather, temperature in zip (days, weathers, temperatures):
    print(f"On {day} it was {weather} and the temperature was {temperature} degree celsius.")

# Or equivalently:
# for t in zip (days, weathers, temperatures):
#   print("On {} it was {} and the temperature was {} degrees celsius.".format(*t))

On Monday it was rainy and the temperature was 10 degree celsius.
On Tuesday it was rainy and the temperature was 12 degree celsius.
On Wednesday it was sunny and the temperature was 12 degree celsius.
On Thursday it was cloudy and the temperature was 9 degree celsius.
On Friday it was rainy and the temperature was 9 degree celsius.
On Saturday it was sunny and the temperature was 11 degree celsius.
On Sunday it was sunny and the temperature was 11 degree celsius.


## Exercise 1.11 (interleave)

- Create a function named interleave.
- The function takes an arbitrary number of lists as parameters.
- The input lists are assumed to have equal length.
- The function should return a single list with elements interleaved.
- Test the interleave function from the main program.
- interleave([1,2,3], [20,30,40], ['a', 'b', 'c']) 
- should return [1, 20, 'a', 2, 30, 'b', 3, 40, 'c']
- Use the zip function to implement interleave. Remember the extend method of list objects.

In [25]:
def interleave(*datas):
    result= []
    # we iterate through each element
    # *datas inside zip is turn multiple list into one list
    for data in zip(*datas):
        # we can extend each element in one list
        result.extend(data)    

    return result
def main():
    print(interleave([1, 2, 3], [20, 30, 40], ['a', 'b', 'c']))
main()

[1, 20, 'a', 2, 30, 'b', 3, 40, 'c']


## Enumerating sequences
Iterating with Indices in Python:
Python simplifies iteration through elements using for loops, but when you need to access the index of the current element, you can use Python's enumerate function. Here's an example where we find the second occurrence of the integer 5 in a list.
- The enumerate(L) function call can be thought to be equivalent to zip(range(len(L)), L).

In [26]:
L = [1, 2, 98, 5, -1, 2, 0, 5, 10]
counter = 0
for i, x, in enumerate(L):
    if x == 5:
        counter += 1
        if counter == 2:
            break
print(i)

7


## Dictionaries in Python:
- Dictionaries are dynamic and unordered containers.  
- Each (key, value) pair is called an item in the dictionary.
- A dynamic, unordered container that uses keys to access value-s. 
- Create it with comma-separated key-value pairs enclosed in braces. 
- Keys and values are separated by a colon, forming (key, value) items. 
- Keys can have different types, as long as they are hashable.

In [27]:
d = {"key1":"value1", "key2":"value2"}
print(d["key1"])
print(d["key2"])

value1
value2


In [28]:
# Alternative Syntax
dict([("key1", "value1"), ("key2", "value2"), ("key3", "value3")])
dict(key1="value1", key2="value2", key3="value3")


{'key1': 'value1', 'key2': 'value2', 'key3': 'value3'}

In [29]:
# Heading Non-Existing Keys:
d = {}
d[2] = "value"

## Dictionary Methods:

- d.copy()
- d.items()
- d.keys()
- d.values()
- d.get(k[,x])
- d.clear()
- d.update(d1)
- d.setdefault(k[,x])
- d.pop(k[,x])
- d.popitem()

## Sets in Python:
A dynamic, unordered container storing unique keys, and they must be hashable.

In [33]:
s = {1, 1, 1}
print(s)
s = set([1, 2, 2, 'a'])
print(s)
s = set() # empty set
s.add(7) # add one element
print(s)

{1}
{1, 2, 'a'}
{7}


In [31]:
s = "mississippi"
print(f"There are {len(set(s))} distinct characters in {s}")

There are 4 distinct characters in mississippi


- Set Methods:
```
s.copy()
s.issubset(s1)
s.issuperset(s1)
s.union(s1)
s.intersection(s1)
s.difference(s1)
s.symmetric_difference(s1)
```

In [34]:
# Use Operators for Operations:
s = set([1, 2, 7])
t = set([2, 8, 9])
print("Union:", s | t)
print("Intersection:", s & t)
print("Difference:", s - t)
print("Symmetric difference:", s ^ t)

Union: {1, 2, 7, 8, 9}
Intersection: {2}
Difference: {1, 7}
Symmetric difference: {1, 7, 8, 9}


### Mutating Methods:
```
s.add(x)
s.clear()
s.discard(x)
s.pop()
s.remove(x)
```

### Mutating Operatos:
```
|= for union
&= for intersection
-= for difference
^= for symmetric difference
```

## Exercise 1.12 (distinct characters)
- Create distinct_characters function, taking a list of strings as input. 
- Return a dictionary with strings as keys and distinct character counts as values.
- Use the set container to temporarily store the distinct characters in a string. Example of usage:
    - distinct_characters(["check", "look", "try", "pop"]) 
    - should return { "check" : 4, "look" : 3, "try" : 3, "pop" : 2}.

In [36]:
def distinct_characters(L):
    result = {}
    for word in L:
        result[word] = len(set(word))
    return result

def main():
    print(distinct_characters(["check", "look", "try", "pop"]))
main()

{'check': 4, 'look': 3, 'try': 3, 'pop': 2}


## Miscellaneous Python Points:

- Use `in` operator to check for element presence in a container.
- For strings, `in` can be used to check if one is part of another.
- Unpack container elements into variables.
- In dictionaries, keys are used by default in membership testing and unpacking.
- To remove variable bindings, use `del`.
- `del` can be used to delete items from containers and slices.
- Later, we'll see `del` for deleting attributes from objects.

In [1]:
print(1 in [1,2])
d = dict(a=1, b=3)
print("b" in d)
s=set()
print(1 in s)
print("x" in "text")


True
True
False
True


In [2]:
print("issi" in "mississippi")
print("issp" in "missippi")


True
False


In [3]:
first, second = [4,5]
a, b, c = "bye"
print(c)
d=dict(a=1, b=3)
key1, key2 = d
print(key1, key2)

e
a b


In [4]:
for key, value in d.items():
    print(f"For key '{key}' value {value} was stored")

For key 'a' value 1 was stored
For key 'b' value 3 was stored


In [6]:
s = "hello"
del s
# print(s) # This would cause an error

NameError: name 's' is not defined

In [7]:
L=[13, 23, 40, 100]
del L[1]
print(L)

[13, 40, 100]


### Exercise 1.13 (reverse dictionary)

In [18]:
"""
Creating a Finnish-English Dictionary:

Given a dictionary with English words as keys and lists of Finnish words as values.
Create a function called reverse_dictionary.
The function generates a Finnish to English dictionary.
Ensure handling synonyms and homonyms.

Out put we want
{'liikuttaa': ['move'], 'piilottaa': ['hide'], 'salata': ['hide'], 'kuusi': ['six', 'fir']}
"""

def reverse_dictionary(d):
    result = {}
   
    for key, values in d.items():
        for value in values:
          
            # check if current value is in a new result key list
            # if exist we have synonymous
            if value in result.keys():
                result[value].append(key)
            else:
                result[value] = [key]
    return result


def main():
    d={'move': ['liikuttaa'], 'hide': ['piilottaa', 'salata'], 'six': ['kuusi'], 'fir': ['kuusi']}
    result = reverse_dictionary(d)
    print(result)
main()


{'liikuttaa': ['move'], 'piilottaa': ['hide'], 'salata': ['hide'], 'kuusi': ['six', 'fir']}


In [20]:
# Exercise 1.14 (find matching)
"""
Write function find_matching that gets a list of strings and 
a search string as parameters. 
The function should return the indices to 
those elements in the input list that contain the search string. 
Use the function enumerate.
An example: 
find_matching(["sensitive", "engine", "rubbish", "comment"], "en") 
should return the list [0, 1, 3].
"""

def find_matching(L, pattern):
    result = []
    for index, word in enumerate(L):
        if pattern in word:
            result.append(index)
    return result
def main():
    result = find_matching(["sensitive", "engine", "rubbish", "comment"], "en") 
    print(result)
main()

[0, 1, 3]


## Creating Data Structures in Python:

- Use list comprehensions for compact list creation.

    - Example: `L = [a**3 for a in range(1,11)]`
- List comprehensions have the form: `[expression for element in iterable lc-clauses]`.
    - lc-clauses may include for and if clauses.
- Nested list comprehensions for more complex structures.

    - Example: `[100*a + 10*b + c for a in range(0,10) for b in range(0,10) for c in range(0,10) if a <= b <= c]`
- Use generator expressions when iterating once to save memory.
    - Replace brackets with parentheses: `G = (100*a + 10*b + c for a in range(0,10) for b in range(0,10) for c in range(0,10) if a <= b <= c)`
- Dictionary comprehension creates dictionaries.

    - Example: `d = {k: k**2 for k in range(10)}`
- Set comprehension creates sets.
    - Example: `s = {i*j for i in range(10) for j in range(10)}`

In [21]:
L=[ 100*a + 10*b +c for a in range(0,10)
                    for b in range(0,10)
                    for c in range(0,10) 
                    if a <= b <= c]
print(L)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 22, 23, 24, 25, 26, 27, 28, 29, 33, 34, 35, 36, 37, 38, 39, 44, 45, 46, 47, 48, 49, 55, 56, 57, 58, 59, 66, 67, 68, 69, 77, 78, 79, 88, 89, 99, 111, 112, 113, 114, 115, 116, 117, 118, 119, 122, 123, 124, 125, 126, 127, 128, 129, 133, 134, 135, 136, 137, 138, 139, 144, 145, 146, 147, 148, 149, 155, 156, 157, 158, 159, 166, 167, 168, 169, 177, 178, 179, 188, 189, 199, 222, 223, 224, 225, 226, 227, 228, 229, 233, 234, 235, 236, 237, 238, 239, 244, 245, 246, 247, 248, 249, 255, 256, 257, 258, 259, 266, 267, 268, 269, 277, 278, 279, 288, 289, 299, 333, 334, 335, 336, 337, 338, 339, 344, 345, 346, 347, 348, 349, 355, 356, 357, 358, 359, 366, 367, 368, 369, 377, 378, 379, 388, 389, 399, 444, 445, 446, 447, 448, 449, 455, 456, 457, 458, 459, 466, 467, 468, 469, 477, 478, 479, 488, 489, 499, 555, 556, 557, 558, 559, 566, 567, 568, 569, 577, 578, 579, 588, 589, 599, 666, 667, 668, 669, 677, 678, 679, 688, 689, 699, 777, 778, 779,

In [22]:
d={ k : k**2 for k in range(10)}
print(d)

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}


In [23]:
s={ i*j for i in range(10) for j in range(10)}
print(s)

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15, 16, 18, 20, 21, 24, 25, 27, 28, 30, 32, 35, 36, 40, 42, 45, 48, 49, 54, 56, 63, 64, 72, 81}


In [26]:
# Exercise 1.15 (two dice comprehension)
"""
Redo the earlier exercise which printed all the pairs of two dice results that sum to 5. 
But this time use a list comprehension. Print one pair per line.
"""

def main():
    target = 5
    """
    for firstDie in range(1, 7):
        for secondDie in range(1, 7):
            if firstDie + secondDie == target:
                print(f"({firstDie}, {secondDie})")
    """
    
    [print(f"({firstDie}, {secondDie})") for firstDie in range(1, 7) for secondDie in range(1,7) if firstDie + secondDie == target]
main()   

(1, 4)
(2, 3)
(3, 2)
(4, 1)


## Processing Sequences in Python:

- Python supports first-class functions, allowing you to:
    - Pass a function as a parameter to another function.
    - Return a function as a return value from a function.
    - Store a function in a data structure or a variable.
- Map and Lambda Functions:

    - Map function takes a list and a function as parameters, returning a new list with transformed elements.
    - The parameter function must take one value in and return a value out.


In [32]:
# Example using double function:
def double(x):
    return 2*x
L = [12, 4, -1]
print(list(map(double, L)))

[24, 8, -2]


In [33]:
# Converting a map object to a list:
result = list(map(double, L))
print(result)

[24, 8, -2]


In [29]:
# Example: Converting string numbers to integers using map:

s = "12 43 64 6"
L = s.split()
print(L)
print(sum(map(int, L)))

['12', '43', '64', '6']
125


In [30]:
"""
    Lambda expressions create nameless functions for one-time use:
    Example replacing add_double_and_square with a lambda function
"""
L = [2, 3, 5]
print(list(map(lambda x: 2*x + x**2, L)))

[8, 15, 35]


In [35]:
# Exercise 1.16 (transform)
"""
Takes two strings as parameters.
Splits the strings into words.
Converts the words into lists of integers.
Returns a list of integers, which are the products of corresponding positions in the two lists.

For example transform("1 5 3", "2 6 -1") 
should return the list of integers [2, 30, -3].
"""
def transform(s1, s2):
    first = list(map(int, s1.split()))
    second = list(map(int, s2.split()))
    return [x * y for x, y in zip(first, second)]
    #return list(map(lambda x, y: x * y, first, second))

def main():
    s1, s2 = "1 5 3", "2 6 -1"
    result = transform(s1, s2)
    print(result)
main()

[2, 30, -3]
